my_methods.rb

```diff
@@ -0,0 +1,245 @@
+      end
+      index += 1
+    end
+    self
+  end
+
+  def my_select
+    return_arr = []
+    return_hash = {}
+    return to_enum(:my_select) unless block_given?
+
+    if is_a?(Hash)
+      my_each do |key, value|
+        return_hash[key] = value if yield(key, value)
+      end
+      return_hash
+    else
+      my_each do |element|
+        return_arr << element if yield(element)
+      end
+      return_arr
+    end
+  end
+
+  def my_all?(pattern = nil)
+    if pattern.nil?
+      if block_given?
+        my_each do |element|
+          return false unless yield(element)
+        end
+      else
+        my_each do |element|
+          return false if element == false || element.nil?
+        end
+      end
+      true
+    elsif pattern.is_a?(Regexp)
+      my_each do |element|
+        return false unless element.match(pattern)
+      end
+      true
+    elsif pattern.is_a?(Module)
+      my_each do |element|
+        return false unless element.is_a?(pattern)
+      end
+    else
+      my_each do |element|
+        return false unless element === pattern
+      end
+    end
+    true
+  end
+
+  def my_any?(pattern = nil)
+    if pattern.nil?
+      if block_given?
+        my_each do |element|
+          return true if yield(element)
+        end
+      else
+        my_each do |element|
+          return true unless element == false || element.nil?
+        end
+      end
+    elsif pattern.is_a?(Regexp)
+      my_each do |element|
+        return true if element.match(pattern)
+      end
+    elsif pattern.is_a?(Module)
+      my_each do |element|
+        return true if element.is_a?(pattern)
+      end
+    else
+      my_each do |element|
+        return true if element === pattern
+      end
+    end
+    false
+  end
+
+  def my_none?(pattern = nil)
+    if pattern.nil?
+      if block_given?
+        my_each do |element|
+          return false if yield(element)
+        end
+      else
+        my_each do |element|
+          return false if element
+        end
+      end
+    elsif pattern.is_a?(Regexp)
+      my_each do |element|
+        return false if element.match(pattern)
+      end
+    elsif pattern.is_a?(Class)
+      my_each do |element|
+        return false if element.is_a?(pattern)
+      end
+    else
+      my_each do |element|
+        return false if element === pattern
+      end
+    end
+    true
+  end
+
+  def my_count(arg = nil)
+    count = 0
+    if arg.nil?
+      if block_given?
+        my_each do |element|
+          count += 1 if yield(element)
+        end
+        count
+      else
+        size
+      end
+    else
+      my_each do |element|
+        count += 1 if element === arg
+      end
+      count
+    end
+  end
+
+  def my_map
+    return_arr = []
+    if is_a?(Hash)
+      my_each do |key, value|
+        return_arr << yield(key, value)
+      end
+    else
+      my_each do |element|
+        return_arr << yield(element)
+      end
+    end
+    return_arr
+  end
```

Comment:
`my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://github.com/TheOdinProject/curriculum/blob/master/ruby_programming/archive/basic_ruby/project_advanced_building_blocks.md#project-2-enumerable-methods))
- returns an Enumerator if no block is given.
___
Now the updated Part,
![image](https://user-images.githubusercontent.com/38625956/82996239-7fb52200-a026-11ea-8069-55ef52e6439a.png)
- executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
```ruby
my_proc = Proc.new { |num| num > 10 }
[11, 2, 3, 15].my_map(my_proc) {|num| num < 10 } # should return [true, false, false, true]
```